pkg/util/util_test.go

```diff
@@ -332,3 +334,383 @@ func TestJoinQualifiedName(t *testing.T) {
+enp0s26u1u2u2	00000000	0100FE0AA	0003	0	0	1024	00000000	0	0	0                                                                   
+enp0s26u1u2u2	0000FE0A	00000000	0001	0	0	0	0080FFFF	0	0	0                                                                      
+docker0	000011AC	00000000	0001	0	0	0	0000FFFF	0	0	0                                                                            
+virbr0	007AA8C0	00000000	0001	0	0	0	00FFFFFF	0	0	0
+`
+
+type addrInterface interface{}
+
+type addrStruct struct{ val string }
+
+func (a addrStruct) Network() string {
+	return a.val
+}
+func (a addrStruct) String() string {
+	return a.val
+}
+
+type testNetworkInterface struct {
+}
+
+func (_ testNetworkInterface) InterfaceByName(intfName string) (*net.Interface, error) {
+	c := net.Interface{Index: 0, MTU: 0, Name: "enp0s26u1u2u2", HardwareAddr: nil, Flags: 1}
+	return &c, nil
+}
+func (_ testNetworkInterface) Addrs(intf *net.Interface) ([]net.Addr, error) {
+	var ifat []net.Addr
+	ifat = []net.Addr{
+		addrStruct{val: "fe80::2f7:6fff:fe6e:2956/64"}, addrStruct{val: "10.254.71.145/17"}}
+	return ifat, nil
+}
+
+type testNetworkInterfaceIPv6 struct {
+}
+
+func (_ testNetworkInterfaceIPv6) InterfaceByName(intfName string) (*net.Interface, error) {
+	c := net.Interface{Index: 0, MTU: 0, Name: "enp0s26u1u2u2", HardwareAddr: nil, Flags: 1}
+	return &c, nil
+}
+func (_ testNetworkInterfaceIPv6) Addrs(intf *net.Interface) ([]net.Addr, error) {
+	var ifat []net.Addr
+	ifat = []net.Addr{addrStruct{val: "fe80::2f7:6fff:fe6e:2956/64"}}
+	return ifat, nil
+}
+
+type testNetworkInterfaceInvalid1 struct {
+}
+
+func (_ testNetworkInterfaceInvalid1) InterfaceByName(intfName string) (*net.Interface, error) {
+	return nil, fmt.Errorf("unable get Interface")
+}
+func (_ testNetworkInterfaceInvalid1) Addrs(intf *net.Interface) ([]net.Addr, error) {
+	return nil, nil
+}
+
+type testNetworkInterfaceInvalid2 struct {
+}
+
+func (_ testNetworkInterfaceInvalid2) InterfaceByName(intfName string) (*net.Interface, error) {
+	c := net.Interface{Index: 0, MTU: 0, Name: "enp0s26u1u2u2", HardwareAddr: nil, Flags: 1}
+	return &c, nil
+}
+func (_ testNetworkInterfaceInvalid2) Addrs(intf *net.Interface) ([]net.Addr, error) {
+	return nil, fmt.Errorf("unable get Addrs")
+}
+
+type testNetworkInterfaceInvalid3 struct {
+}
+
+func (_ testNetworkInterfaceInvalid3) InterfaceByName(intfName string) (*net.Interface, error) {
+	c := net.Interface{Index: 0, MTU: 0, Name: "enp0s26u1u2u2", HardwareAddr: nil, Flags: 1}
+	return &c, nil
+}
+func (_ testNetworkInterfaceInvalid3) Addrs(intf *net.Interface) ([]net.Addr, error) {
+	var ifat []net.Addr
+	ifat = []net.Addr{addrStruct{val: "fe80::2f7:6ffff:fe6e:2956/64"}}
+	return ifat, nil
+}
+
+func TestParseIPInvalid(t *testing.T) {
+	ip, err := parseIP("0000FE0AA")
+	if err == nil {
+		t.Errorf("expected error got : nil")
+	}
+	if ip != nil {
+		t.Errorf("expected nil got : %v ", ip)
+	}
+}
+
+func TestParseIPInvalidIpv4(t *testing.T) {
```

Comment:
The general pattern in go for writing tests is to write a table-driven test, and I think that will make ALL the difference in this test file.  In other words, don't have two functions which each test one case.  Instead do:

```
testCases := []struct {
    case string
    ip string
    success bool
    expected net.IP
}{
    {"empty", "", false, nil},
    {"too short", "AA", false, nil},
    {"too long", "0011223344", false, nil},
    {"invalid", "invalid!", false, nil},
    {"zero", "00000000", true, net.IP{0, 0, 0, 0}},
    {"ffff", "FFFFFFFF", true, net.IP{0xff, 0xff, 0xff, 0xff}},
    {"valid", "12345678", true, net.IP{78, 56, 34, 12}},
}
for _, tc := range testCases {
    ip, err := parseIP(tc.ip)
    // check tc.success, tc.expected, etc.
    // if an error, log tc.case
}
```

This pattern applies all through the file.